electron/syntheticsGenerator.ts

```diff
@@ -0,0 +1,364 @@
+    } else if (action.command) {
+      formatter.add(toAssertCall(pageAlias, action));
+    }
+
+    this.previousContext = actionInContext;
+    return formatter.format();
+  }
+
+  isNewStep(actioninContext: ActionInContext) {
+    const { action, frameUrl } = actioninContext;
+    if (action.name === 'navigate') {
+      return true;
+    } else if (action.name === 'click') {
+      return this.previousContext?.frameUrl === frameUrl && action.signals.length > 0;
+    }
+    return false;
+  }
+
+  generateStepStart(name: string) {
+    this.insideStep = true;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`step(${quote(name)}, async () => {`);
+    return formatter.format();
+  }
+
+  generateStepEnd() {
+    if (!this.insideStep) {
+      return '';
+    }
+    this.insideStep = false;
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    formatter.add(`});`);
+    return formatter.format();
+  }
+
+  generateHeader() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(0);
+    formatter.add(`
+      const { journey, step, expect } = require('@elastic/synthetics');
+
+      journey('Recorded journey', async ({ page, context }) => {`);
+    return formatter.format();
+  }
+
+  generateFooter() {
+    return `});`;
+  }
+
+  /**
+   * Generates JavaScript code from a custom set of steps and nested actions.
+   *
+   * This function makes no assumptions about where steps should be created,
+   * and instead follows the step definitions the caller has defined.
+   * @param steps IR to use for code generation
+   * @returns a list of the code strings outputted by the generator
+   */
+  generateFromSteps(steps: Steps): string {
+    const text = [];
+    if (this.isSuite) {
+      text.push(this.generateHeader());
+    }
+    this.varsToHoist = this.findVarsToHoist(steps);
+    text.push(this.generateHoistedVars());
+    for (const step of steps) {
+      if (step.actions.length === 0) throw Error('Cannot process an empty step');
+      const name = step.name ?? step.actions[0].title ?? actionTitle(step.actions[0].action);
+      text.push(this.generateStepStart(name ?? ''));
+
+      for (const action of step.actions) {
+        const actionText = this.generateAction(action);
+        if (actionText.length) text.push(actionText);
+      }
+
+      text.push(this.generateStepEnd());
+    }
+    if (this.isSuite) {
+      text.push(this.generateFooter());
+    }
+
+    return text.filter(s => !!s).join('\n');
+  }
+
+  generateHoistedVars() {
+    const formatter = new PlaywrightGenerator.JavaScriptFormatter(this.getDefaultFormatterOffset());
+    for (const varName of this.varsToHoist) {
+      formatter.add(`let ${varName};`);
+    }
+    return formatter.format();
+  }
+
+  isVarHoisted(varName: string) {
+    return this.varsToHoist.indexOf(varName) >= 0;
+  }
+
+  getDefaultFormatterOffset() {
```

Comment:
nit: getDefaultOffset, formatter is wordy as we are already in the formatter.

Reply:
https://github.com/elastic/synthetics-recorder/pull/211/commits/de8dfafdee1899f0c71b20746e46b0a92932ef8f